account-kit/smart-contracts/src/ma-v2/account/semiModularAccountV2.ts

```diff
@@ -20,18 +21,53 @@ import {
   type Hex,
   type Transport,
   hexToBigInt,
+  zeroAddress,
 } from "viem";
 import { accountFactoryAbi } from "../abis/accountFactoryAbi.js";
 import { getDefaultMAV2FactoryAddress } from "../utils.js";
 import { standardExecutor } from "../../msca/account/standardExecutor.js";
 import { singleSignerMessageSigner } from "../modules/single-signer-validation/signer.js";
 import { InvalidEntityIdError, InvalidNonceKeyError } from "@aa-sdk/core";
+import { modularAccountAbi } from "../abis/modularAccountAbi.js";
+import { serializeModuleEntity } from "../actions/common/utils.js";
+
+const executeUserOpSelector: Hex = "0x8DD7712F";
 
 export const DEFAULT_OWNER_ENTITY_ID = 0;
 
+export type CalldataEncoder = {
+  encodeCallData: (callData: Hex) => Promise<Hex>;
+};
+
+export type ExecutionDataView = {
+  module: Address;
+  skipRuntimeValidation: boolean;
+  allowGlobalValidation: boolean;
+  executionHooks: readonly Hex[];
+};
+
+export type ValidationDataView = {
+  validationHooks: readonly Hex[];
```

Comment:
Especially, I think marking arrays as `readonly` isn't very helpful, because this doesn't stop the contents of the array from changing; it only stops the variable from being reassigned.